lib/block_view_association.go

```diff
@@ -731,32 +731,86 @@ func (bav *UtxoView) GetUserAssociationsByAttributes(associationQuery *UserAssoc
+	// Store matching associations from the db + UTXO view in a slice.
+	var associationEntries []*UserAssociationEntry
+	// First, pull matching association entries from the UTXO view so that
+	// we can track deleted association entries and can properly limit the
+	// number of entries retrieved from the database.
+	newUtxoAssociationEntries, deletedUtxoAssociationIDs := bav._getUtxoUserAssociationEntriesByAttributes(associationQuery)
+	// Store a map of UTXO view association entries to prevent duplicates pulled from the db.
+	utxoAssociationEntryMap := make(map[*BlockHash]*UserAssociationEntry)
+	for _, associationEntry := range newUtxoAssociationEntries {
+		utxoAssociationEntryMap[associationEntry.AssociationID] = associationEntry
+	}
+	// Check the database for matching association entries.
+	dbAssociationEntries, prefixType, err := bav.GetDbAdapter().GetUserAssociationsByAttributes(associationQuery, deletedUtxoAssociationIDs)
 	if err != nil {
-		return nil, err
-	}
-	for _, dbAssociationEntry := range dbAssociationEntries {
-		associationEntryMap[dbAssociationEntry.AssociationID] = dbAssociationEntry
-	}
-	// Next, check the UTXO view for most recent association entries which
-	// take priority over any with the same AssociationID from the database.
-	newUtxoAssociationEntries, deletedUtxoAssociationEntryIDs := bav._getUtxoUserAssociationEntriesByAttributes(associationQuery)
-	for _, utxoAssociationEntry := range newUtxoAssociationEntries {
-		// Add new association entries found in the UTXO view.
-		associationEntryMap[utxoAssociationEntry.AssociationID] = utxoAssociationEntry
+		return nil, errors.Wrapf(err, "GetUserAssociationsByAttributes: ")
 	}
-	for utxoAssociationEntryID := range deletedUtxoAssociationEntryIDs {
-		// Remove deleted association entries found in the UTXO view.
-		delete(associationEntryMap, utxoAssociationEntryID)
+	// Sort the UTXO association entries according to the query's specified sort order.
+	sortedUtxoAssociationEntries, err := bav.GetDbAdapter().SortUserAssociationEntriesByPrefix(
+		newUtxoAssociationEntries, prefixType, associationQuery.SortDescending,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "GetUserAssociationsByAttributes: ")
+	}
+	// Skip UTXO association entries up to the last seen association ID if specified.
+	var utxoAssociationEntries []*UserAssociationEntry
+	isNewEntry := associationQuery.LastSeenAssociationID == nil
+	for _, associationEntry := range sortedUtxoAssociationEntries {
+		if associationEntry.AssociationID.IsEqual(associationQuery.LastSeenAssociationID) {
+			isNewEntry = true
+			continue
+		}
+		if !isNewEntry {
+			continue
+		}
+		utxoAssociationEntries = append(utxoAssociationEntries, associationEntry)
 	}
-	// Convert map to slice.
-	var associationEntries []*UserAssociationEntry
-	for _, matchingAssociationEntry := range associationEntryMap {
-		associationEntries = append(associationEntries, matchingAssociationEntry)
+	if associationQuery.SortDescending {
```

Comment:
If we sort ascending, we pull from the database first and the view second (deleting any entries that are deleted in the view). If we sort descending, we pull from the UTXO view first and the database second. So we can't do an in-memory sort.

Reply:
well you're not guaranteed to have entries in the view be "newer" than entries in the DB, does this change how we should sort?

Reply:
IMO no. It feels like we just need one to return db then UTXO view and the other return UTXO view then db. Otherwise, it's weird for an end-user to query order ASC and then query order DESC and see the same entries from the UTXO view first both times for example.

Also importantly, sort ASC|DESC is not necessarily by age. It's by db prefix key for badger and by AssociationID (BlockHash) for postgres. The UTXO view follows the sorting pattern of the database.

Reply:
well why can't we just append all the entries from the view and the db into one slice and then apply the sort criteria and then take the slice of elements we want? I was thinking this would actually get us the most consistent set of results, adhering to the sorting order defined by the db? I think you could replace lines 827 to 873 with the snippet below.
```
	allAssociationEntries := utxoAssociationEntries
	for _, dbEntry := range dbAssociationEntries {
		if _, exists := utxoAssociationEntryMap[dbEntry.AssociationID]; !exists {
			allAssociationEntries = append(allAssociationEntries, dbEntry)
		}
	}
	allAssociationEntriesSorted, err := bav.GetDbAdapter().SortUserAssociationEntriesByPrefix(
		allAssociationEntries, prefixType, associationQuery.SortDescending)
	maxIndex := associationQuery.Limit
	if maxIndex > len(allAssociationEntriesSorted) {
		maxIndex = len(allAssociationEntriesSorted)
	}
	return allAssociationEntries[:maxIndex], nil
```

Reply:
you could do some sort of binary search insertion, but I think that's unnecessary rn instead of using the helper function you defined in the db adapter.